System/IO/Temp.hs

```diff
@@ -212,3 +217,49 @@ mkPrivateDir s = System.Posix.createDirectory s 0o700
 -- "/home/feuerbach"
 getCanonicalTemporaryDirectory :: IO FilePath
 getCanonicalTemporaryDirectory = getTemporaryDirectory >>= canonicalizePath
+
+
+-- | Make a temporary file or directory name in the given directory, similar to unix 'mktemp'.
+-- This does not create a file.
+--
+-- Note that this might be subject to race conditions. Only use this on secure directories.
+mkTempFileName :: FilePath   -- ^ directory in which to test for available random filenames
+               -> String     -- ^ File name template. If the template is \"foo.ext\" then
+                             -- the created file will be \"fooXXX.ext\" where XXX is some
+                             -- random number. Note that this should not contain any path
+                             -- separator characters.
+               -> IO FilePath
+mkTempFileName tmp_dir template
```

Comment:
This gives filenames with predictable filenames.

Even if you're in a "secure" directory (please document what this means here), there's a problem.

Consider a program that starts up, makes 10 temporary file in some secure directory with some content then exits.

If you re-run this same program a bunch, you'll eventually likely to get the same PID and the program fails.

So this has the same issue as mktemp(): race condition between file test and potentially opening a file and predictable file names. Actually it's worse, many mktemp() implementations use PID<letter> as the replacement for the pattern which gives 26 possible names while this gives at most 10.

It seems that even just generating random filename is better than generating many deterministic (within the process) ones and trying each one.

That still has the race condition of course.

Reply:
> This gives predicable filenames.
> 
> Even if you're in a "secure" directory (please document what this means here), there's a problem.
> 

> Consider a program that starts up, makes 10 temporary file in some secure directory with some content then exits.
> 
> If you re-run this same program a bunch, you'll eventually likely to get the same PID and the program fails.
> 

That's incorrect. The implementation behaves like mkdtemp and retries

> So this has the same issue as mktemp(): race condition between file test and potentially opening a file and predictable file names. Actually it's worse, many mktemp() implementations use PID<letter> as the replacement for the pattern which gives 26 possible names while this gives at most 10 (EDIT: At most 10 per `mkTempFileName` invocation; due to counter, subsequent invocations give 10 other names which is different from `mktemp` AFAIU; still...).
> 
> It seems that even just generating random filename is better than generating many deterministic (within the process) ones and trying each one.
> 
> That still has the race condition of course.

Well, this is basically the implementation that's already used by this library: https://hackage.haskell.org/package/base-4.15.0.0/docs/src/System-IO.html#openTempFile

And of course, this is meant for **temporary** files, not persistent files.

So the current implementation has exactly the same behavior, so you could file a bug against `base`.

I'd most certainty adjust the code if a patch was accepted in base.

Reply:
>So the current implementation has exactly the same behavior, so you could file a bug against base.

No, not at all.

The difference between `openTempFile` and what you have here is this:

```
-- With some exceptions (see below), the file will be created securely
-- in the sense that an attacker should not be able to cause
-- openTempFile to overwrite another file on the filesystem using your
-- credentials, by putting symbolic links (on Unix) in the place where
-- the temporary file is to be created.  On Unix the @O_CREAT@ and
-- @O_EXCL@ flags are used to prevent this attack, but note that
-- @O_EXCL@ is sometimes not supported on NFS filesystems, so if you
-- rely on this behaviour it is best to use local filesystems only.
```

What you're proposing here just rolls a dice and gives you a random filename that it randomly checked for existence of on the filesystem. What's in base is easy to use as part of programs without some strange unintended security hole. What you're proposing is adding something inherently broken to a very popular library with bunch of caveats in the docs.

Reply:
You didn't really read what I said.

The function in base has the same randomization trick. So if you create a million files and forget to clean them up, there's a chance the next process will hang, because it's iterating through a million already existing files. 

Wrt the security thing: WE ALREADY ESTABLISHED THAT IT'S ONLY REASONABLE IN A SECURE DIRECTORY.